x-pack/qa/security-example-spi-extension/src/javaRestTest/java/org/elasticsearch/example/realm/CustomRealmIT.java

```diff
@@ -23,12 +23,16 @@
  */
 public class CustomRealmIT extends ESRestTestCase {
 
+    // These are configured in build.gradle
+    public static final String USERNAME= "test_user";
+    public static final String PASSWORD = "secret_password";
+
     @Override
     protected Settings restClientSettings() {
         return Settings.builder()
-                .put(ThreadContext.PREFIX + "." + CustomRealm.USER_HEADER, CustomRealm.KNOWN_USER)
-                .put(ThreadContext.PREFIX + "." + CustomRealm.PW_HEADER, CustomRealm.KNOWN_PW.toString())
-                .build();
+            .put(ThreadContext.PREFIX + "." + CustomRealm.USER_HEADER, USERNAME)
+            .put(ThreadContext.PREFIX + "." + CustomRealm.PW_HEADER, PASSWORD)
```

Comment:
This is rather a question for my learning. Why are these needed here? This configures the threadContext to always contain the right headers of username and password. Does that mean transport client can always connect without providing credentials?

Reply:
This is for rest client not transport client, but it follow the same patterns as transport client (so it's a bit confusing).

In `ESRestTestCase`, rather than configure a client in code, the pattern is to configure the headers in a `Settings` object that is used to configure the client.

This is setting the username & password that the default client uses so that it can connect to the external cluster.